### PR TITLE
PCT-1482: QAN manager refactor + instance manager

### PR DESCRIPTION
**Pull request by mpaolino**
This branch includes the refactor on instance manager and qan manager (this branch was branched of PCT-1481-instances-tree, which was branched of release). Several other components that depend on instance manager and instances objects still need to be refactored to use new code and to pass their tests, that is part of next task (PCT-1483). You'll notice that in some places there are still references to "instanceId" instead of "UUID", this is just to be able to compile and run qan manager and instance manager tests, those are going to be changed asap. Also, final Rest endpoints for instances will be updated before full refactor is completed.

New behavior for instances and qan manager:

- Instances manager will get the complete instance tree via WS command or it will get it consuming a webservice. Updates to this tree will be pushed from API using WS command ("UpdateTree" cmd for instance service, proto.InstanceSync as payload). Updates of MySQL instance properties will pushed to API with a PUT on the corresponding instance endpoint but leaving subsystems out of it (they are ignored by the webservice). Full instance tree will be requested from API during qan manager initialization if no config file was found. This tree will only be the instances of the OS system where the agent is running.
- Add/delete/updates to instances will be passed along by the API with the tree and its version
- All instances will be stored in a single instances.conf file in config directory.
- No more single qan.conf file, QAN manager will keep separated configs for every configured instance, with naming qan-<uuid>.conf

I know this is hard to follow but I need some feedback from devs to see if they are ok on the direction taken, I'd be glad to answer your questions.

### CODE REVIEW

File: Godeps.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
 		},
 		{
 			"Pkg": "github.com/percona/cloud-protocol",
-			"Rev": "d729c25e20ee8fec56179358066028f48131468b"
+			"Rev": "1ca0831678f0d3a41b32bea1af2013a97aa9ad7a"
```
Comment (mpaolino) [inline Godeps.json]: This is @cfsalguero cloud-protocol updated for PCTv3.

File: instance/manager.go [hunk 2]
```diff
@@ -110,80 +108,106 @@ func (m *Manager) Stop() error {
...
 	}
 
 	switch cmd.Cmd {
-	case "Add":
-		err := m.repo.Add(it.Service, it.InstanceId, it.Instance, true) // true = write to disk
+	case "Update":
+		added := make([]proto.Instance, 0)
+		deleted := make([]proto.Instance, 0)
+		updated := make([]proto.Instance, 0)
+		err := m.repo.UpdateTree(*it, &added, &deleted, &updated, true) // true = write to disk
```
Comment (daniel-nichter) [inline instance/manager.go]: Passing `*it` is probably not necessary.

Style-wise, I don't think any other code passes in args that get set/populated. Imo, better approach is to return a data struct as a return value. This makes testing cleaner:
```go
gotChanges, err := repo.UpdateTree(tree)
t.Check(gotChanges, DeepEquals, expectChanges)
```

File: instance/repo.go [hunk 3]
```diff
@@ -18,257 +18,379 @@
...
-	configDir string
-	api       pct.APIConnector
-	// --
-	it  map[string]interface{}
-	mux *sync.RWMutex
+	logger         *pct.Logger
+	configDir      string
+	api            pct.APIConnector
+	it             map[string]*proto.Instance
+	tree           *proto.Instance
```
Comment (daniel-nichter) [inline instance/repo.go]: Perhaps best to make these just `proto.Instance` to communicate and enforce that agent doesn't change the tree, only the API can do that.

File: instance/repo.go [hunk 4]
```diff
@@ -18,257 +18,379 @@
...
 }
 
-func (r *Repo) Init() error {
-	for service, _ := range proto.ExternalService {
-		if err := r.loadInstances(service); err != nil {
-			return fmt.Errorf("%s: %s", service, err)
+// Determine if two instances are "equal".
+// Instance equality is defined by the equality of all its attributes, plus
+// the number of subsystems and their UUIDs.
+func equalInstances(inst1, inst2 *proto.Instance) bool {
```
Comment (daniel-nichter) [inline instance/repo.go]: I think this is just as simple as `if instance1.UUID == instance2.UUID { // they're equal }`. Is there a case/reason why this isn't sufficient?

File: instance/repo.go [hunk 5]
```diff
@@ -18,257 +18,379 @@
...
+	if inst1.ParentUUID != inst2.ParentUUID ||
+		inst1.UUID != inst2.UUID ||
+		inst1.Name != inst2.Name ||
+		inst1.Created != inst2.Created ||
+		inst1.Deleted != inst2.Deleted ||
+		len(inst1.Subsystems) != len(inst2.Subsystems) ||
+		!reflect.DeepEqual(inst1.Properties, inst2.Properties) {
```
Comment (daniel-nichter) [inline instance/repo.go]: Imho any time `reflect` shows up in normal code, it's a signal that there's a design problem. :-) There is one case we've used it, but there's a big code comment there about why it's necessary.

File: instance/repo.go [hunk 6]
```diff
@@ -18,257 +18,379 @@
...
+	r.tree = newTree
+	return r.updateInstanceIndex()
 }
 
-func (r *Repo) Add(service string, id uint, data []byte, writeToDisk bool) error {
-	r.logger.Debug("Add:call")
-	defer r.logger.Debug("Add:return")
+// Downloads and returns the instances tree data from API.
+func (r *Repo) downloadInstances() (data []byte, err error) {
+	errors.New(fmt.Sprintf("Downloading instance config file from API"))
```
Comment (daniel-nichter) [inline instance/repo.go]: What's this line doing?

File: instance/repo.go [hunk 7]
```diff
@@ -18,257 +18,379 @@
...
+	url := r.api.EntryLink("insts")
```
Comment (daniel-nichter) [inline instance/repo.go]: We'll need to define this properly in API since it's a very important entry link. I'd call it "system_tree" (I think we're using that style for link relations?).

File: instance/repo.go [hunk 8]
```diff
@@ -18,257 +18,379 @@
...
+	for uuid, _ := range oldIt {
+		// Does it exist in new tree?
+		if _, ok := r.it[uuid]; ok {
+			// Is it the same as the former instance?
+			// We use custom compare method instead of using DeepEquals
+			// on the instances as they include references to child instances;
+			// a change in a child instance means DeepEquals will detect that
+			// a parent was also modified.
+			if !equalInstances(oldIt[uuid], r.it[uuid]) {
+				*updated = append(*updated, *(r.it[uuid]))
```
Comment (daniel-nichter) [inline instance/repo.go]: Use of dereference signals a problem to me, especially like ` *(r.it[uuid])`.

File: test/gocheck.go [hunk 9]
```diff
@@ -73,7 +74,7 @@ func IsDeeply(got, expect interface{}) (result bool, error string) {
 }
 
 func deepEquals(got reflect.Value, expect reflect.Value, level uint) (bool, string) {
```
Comment (daniel-nichter) [inline test/gocheck.go]: Remove all references to this file and use `. "github.com/go-test/test"` instead for `IsDeeply()` and `Dump()`.

File: instance/instance_test.go [hunk 10]
```diff
@@ -84,110 +99,150 @@ func (s *RepoTestSuite) TearDownSuite(t *C) {
...
 
-	// Instance ID must be > 0.
-	err = im.Add("mysql", 0, data, false)
-	t.Assert(err, NotNil)
+	// Request an instance tree copy (using instances-1.conf fixture)
+	tree := s.im.Instances()
+
+	// Lets modify one instance in our test tree copy
+	// index 1 corresponds to instance c540346a644b404a9d2ae006122fc5a2
+	tree.Subsystems[1].Properties["dsn"] = "other DSN"
```
Comment (daniel-nichter) [inline instance/instance_test.go]: DSN is no longer a property, it's a first-class attribute of the instance. This was recently updated in the spec.

File: instance/manager.go [hunk 11]
```diff
@@ -110,80 +108,106 @@ func (m *Manager) Stop() error {
 	return nil
 }
 
+/////////////////////////////////////////////////////////////////////////////
+// Implementation
+/////////////////////////////////////////////////////////////////////////////
+
+// Adds a MySQL instance to MRM
+func (m *Manager) mrmMySQL(inst proto.Instance) error {
```
Comment (daniel-nichter) [inline instance/manager.go]: Function name is not very descriptive.

File: instance/manager.go [hunk 12]
```diff
@@ -110,80 +108,106 @@ func (m *Manager) Stop() error {
...
+	itDSN, ok := inst.Properties["dns"]
```
Comment (daniel-nichter) [inline instance/manager.go]: Just `dsn` is sufficient as a variable name. :-) It implies an underlying "it" (instance).

File: instance/manager.go [hunk 13]
```diff
@@ -110,80 +108,106 @@ func (m *Manager) Stop() error {
...
+	if !ok {
+		return errors.New("Missing DSN in MySQL instance " + inst.UUID)
+	}
+	ch, err := m.mrm.Add(itDSN)
+	if err != nil {
+		return err
+	}
+	m.mrmChans[itDSN] = ch
```
Comment (daniel-nichter) [inline instance/manager.go]: Why is the instance manager keep track of MRMS channels? The MRMS service should do that internally. I know this is existing code, but it seems to violate a separation of concerns.

File: instance/manager.go [hunk 14]
```diff
@@ -110,80 +108,106 @@ func (m *Manager) Stop() error {
...
+	m.status.Update("instance", "Updating info "+safeDSN)
+	err = m.pushInstanceInfo(inst)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Auxiliary function to add MySQL instances on MRM
+func (m *Manager) mrmAddedMySQL(added []proto.Instance) {
```
Comment (daniel-nichter) [inline instance/manager.go]: I suggest present test since it's imperative, e.g. `mrmAddMySQL`. When I first saw "added" I thought there was some kind of post-add operations being done in this func, e.g. "we've added MySQL to MRMS, now (in this func) notify the user" (or something like that).

File: instance/manager.go [hunk 15]
```diff
@@ -110,80 +108,106 @@ func (m *Manager) Stop() error {
...
 }
 
 func (m *Manager) Status() map[string]string {
-	m.status.Update("instance-repo", strings.Join(m.repo.List(), " "))
+	list := m.repo.List()
+	uuids := make([]string, len(list))
+	for _, it := range list {
+		uuids = append(uuids, it.UUID)
+	}
+	m.status.Update("instance-repo", strings.Join(uuids, " "))
```
Comment (daniel-nichter) [inline instance/manager.go]: Just something to think about: this probably isn't going to "scale" for this status line because UUID are long. We'll need some better way to represent whatever real-time status info is important.

File: instance/repo.go [hunk 16]
```diff
@@ -18,257 +18,379 @@
...
-	api       pct.APIConnector
-	// --
-	it  map[string]interface{}
-	mux *sync.RWMutex
+	logger         *pct.Logger
+	configDir      string
+	api            pct.APIConnector
+	it             map[string]*proto.Instance
+	tree           *proto.Instance
+	downloadedInst []byte
```
Comment (daniel-nichter) [inline instance/repo.go]: Dead code.

File: instance/repo.go [hunk 17]
```diff
@@ -18,257 +18,379 @@
...
+	mux            *sync.RWMutex
 }
 
+const (
+	INSTANCES_FILE     = "instances.conf" // relative to Repo.configDir
```
Comment (daniel-nichter) [inline instance/repo.go]: I'd call this file "system-tree.json" because it contains the system tree. I don't think it's a config file?

File: instance/repo.go [hunk 18]
```diff
@@ -18,257 +18,379 @@
...
+	}
+	return true
 }
 
-func (r *Repo) loadInstances(service string) error {
-	files, err := filepath.Glob(r.configDir + "/" + service + "-*.conf")
-	if err != nil {
-		return err
+// Checks if it instance has Prefix == itPrefix
+func hasPrefix(it proto.Instance, itPrefix string) bool {
```
Comment (daniel-nichter) [inline instance/repo.go]: This func is probably overkill. Where it's called, I think just `return it.Prefix == OS_PREFIX` will work?

File: instance/repo.go [hunk 19]
```diff
@@ -18,257 +18,379 @@
...
+		return errors.New("Tree instance root is not of 'OS' type and 'os' prefix")
 	}
 
-	if !valid(service, id) {
-		return pct.InvalidServiceInstanceError{Service: service, Id: id}
+	oldIt := r.it
+	// We need to deep copy the provided tree as it keeps references to
+	// proto.Instances in Subsystems slice that are not copied, hence the caller
+	// can modify them without our knowledge
+	var newTree *proto.Instance
```
Comment (daniel-nichter) [inline instance/repo.go]: Explanation in code comments is good, helpful. If that's a problem, can't we just use an immutable struct, i.e. no pointers? I realize we sometimes use pointers in structs so we can `json:-,omitempty`, but if those are few, then we can just trust callers and the rest of the code rather than do deep copy which, imho, introduces more opportunities for problems and bugs than it solves.

File: instance/manager.go (hunk 14 above)
Comment (mpaolino) [inline instance/manager.go]: yep, makes sense, thanks

File: instance/manager.go (hunk 2 above)
Comment (mpaolino) [inline instance/manager.go]: I guess a single struct with added/updated/deleted could work just fine, I'll update

File: instance/instance_test.go (hunk 10 above)
Comment (mpaolino) [inline instance/instance_test.go]: Oh I see, I like the change, will update to match.

File: instance/manager.go (hunk 11 above)
Comment (mpaolino) [inline instance/manager.go]: You're right it is not, will update.

File: instance/manager.go (hunk 12 above)
Comment (mpaolino) [inline instance/manager.go]: yes

File: instance/manager.go (hunk 13 above)
Comment (mpaolino) [inline instance/manager.go]: I'll dig a little more, to see if I can remove that.

File: instance/manager.go (hunk 15 above)
Comment (mpaolino) [inline instance/manager.go]: I'll think about this, @daniel-nichter: are status lines only meant to be read by users?

File: instance/repo.go (hunk 3 above)
Comment (mpaolino) [inline instance/repo.go]: hmm, I agree that the agent should not change the relationships of the instance tree, but some instance properties are updated by the agent, but in any case, I don't see how not using pointers internally in the repo enforces that. Let me try to explian why are those pointers there. ```tree``` keeps the instance tree, while ```it``` is just and UUID-index to tree elements. I don't want to keep two separated copies I need one set of instances. tree is a pointer because it makes very easy to determine when ```tree``` was not set, which makes easier to detect when the repo was not properly initialized. The reason for using a map to pointers in ```it``` is because I just need an index, working within the repo with pointers is idiomatic, but is also efficient, otherwise I'll be making copies everytime I want to pass something around and that spells trouble with nested structs. See deep copy comment bellow.

File: instance/repo.go (hunk 16 above)
Comment (mpaolino) [inline instance/repo.go]: yep, hopefully go vet + go lint are now going to help me with that.

File: instance/repo.go (hunk 17 above)
Comment (mpaolino) [inline instance/repo.go]: To be honest, I'd naturally search for a ".conf" for this, but I don't have a preference. I'm fine with "system-tree.json" but maybe "instance-tree.json" is more descriptive since we are not storing the system tree but just a partial tree of instances? ("system" is a more broader concept, I think)

File: instance/repo.go (hunk 4 above)
Comment (mpaolino) [inline instance/repo.go]: This is used to determine if an instance was updated by API. UUID does not (normally) change, but other attributes may.

File: instance/repo.go (hunk 5 above)
Comment (mpaolino) [inline instance/repo.go]: Well Properties is dynamic, so I cannot exactly know what properties to compare. I can whoever compare inst1.Properties against inst2.Properties (and the other way around) to see if all their elements match and avoid using DeepEquals, but at the end I'd be doing more or less the same work as DeepEquals.

File: instance/repo.go (hunk 18 above)
Comment (mpaolino) [inline instance/repo.go]: Yes, this initially included checks for Type, so it made a little more sense at the time, not anymore.

File: instance/repo.go (hunk 6 above)
Comment (mpaolino) [inline instance/repo.go]: this should be a log line, fixed

File: instance/repo.go (hunk 7 above)
Comment (mpaolino) [inline instance/repo.go]: Yes we need to, @daniel-nichter what do you think about using "instance_tree" as a little more descriptive name? In our context "system" is a more broad concept IMHO.

File: instance/repo.go (hunk 8 above)
Comment (mpaolino) [inline instance/repo.go]: ```updated``` was a slice passed by reference that was getting mutated, that will change for a returned value, so this is deprecated as it is. The reason to use pointers on ```r.it``` is that I don't want to have two copies of instances, I just need an internal index to the same instances referenced in ```r.tree```, and since I want to return a copy of them I need to dereference.

File: instance/repo.go (hunk 19 above)
Comment (mpaolino) [inline instance/repo.go]: IMHO we need to make an exception here. Nested structs that contain a slice of the same struct type that are passed by value works in an unexpected way. The contained slice is copied but the references to its elements are kept, so you have a inmutable "head" instance with mutable subsystems. For me this spells trouble, since instances will be passed all around the app and they will effectively have global shared instance objects. Following separation of concerns principle, the repo needs to own the instances representations, and return copies to the callers, also should not trust the caller to leave the instance unchanged once it was passed to the repo or received from it.  Example of this behaviour: http://play.golang.org/p/GeUKlwZeOp

File: instance/manager.go (hunk 15 above)
Comment (daniel-nichter) [inline instance/manager.go]: Yeah, basically so. Off and on I've thought maybe they should be somewhat (or completely) machine-readable, but right now they're human-readable. The reason for making them machine-readable is to let API or 3rd-party code programmatically check status of things, but this is probably better done via other means. Maybe in future agent can have 2 status commands: one for humans and machines. In any case, for now, status lines are mean for humans to read.

File: instance/repo.go (hunk 5 above)
Comment (daniel-nichter) [inline instance/repo.go]: I see there's a larger issue in question here re updates to the system tree. I'll send an email, else these comments will become very long.

File: test/gocheck.go (hunk 9 above)
Comment (mpaolino) [inline test/gocheck.go]: IsDeeply and Dump need to be removed from gocheck.go and app.go but a lot of other code depend on them right now. Change should be minimal so I'll do that once this refactor of all components using qan and instance manager are done, code for qan manager and instance manager uses go-test now.

**Comment by mpaolino** [comment]
@daniel-nichter can you have a second look? I'd like to move forward with the services that use this qan/instance manager refactors but only when this is correct. I'll ask FE guys how status updates can be better presented to the user, that will probably influence the way we build and send these messages.

File: instance/manager.go (hunk 13 above)
Comment (mpaolino) [inline instance/manager.go]: Update: this cannot be removed for now. This will be inmediatly addressed with the mrms refactor because right now m.mrm.Remove needs to get the subscribed channel and lots of code use this method (really weird).

File: instance/manager.go [hunk 20]
```diff
@@ -110,80 +107,124 @@ func (m *Manager) Stop() error {
...
-	it := &proto.ServiceInstance{}
-	if err := json.Unmarshal(cmd.Data, it); err != nil {
-		return cmd.Reply(nil, err)
-	}
-
 	switch cmd.Cmd {
-	case "Add":
-		err := m.repo.Add(it.Service, it.InstanceId, it.Instance, true) // true = write to disk
-		if err != nil {
+	case "UpdateTree":
```
Comment (daniel-nichter) [inline instance/manager.go]: Let's keep terminology consistent: `UpdateSystemTree` (i.e. is "system tree" not "tree").

File: instance/manager.go [hunk 21]
```diff
@@ -228,43 +253,27 @@ func GetMySQLInfo(it *proto.MySQLInstance) error {
...
+		&hostname,
+		&distro,
+		&version,
 	)
 	if err != nil {
 		return err
 	}
+	it.Properties["hostname"] = hostname
+	it.Properties["distro"] = distro
+	it.Properties["version"] = version
```
Comment (daniel-nichter) [inline instance/manager.go]: These are ok for now but ultimately they'll go away when `sysconfig` is rewritten to be a full-blown system info collector.

File: instance/manager.go [hunk 22]
```diff
@@ -315,10 +324,12 @@ func (m *Manager) monitorInstancesRestart(ch chan string) {
...
 
-func (m *Manager) pushInstanceInfo(instance *proto.MySQLInstance) error {
-
-	uri := fmt.Sprintf("%s/%s/%d", m.api.EntryLink("instances"), "mysql", instance.Id)
-	data, err := json.Marshal(instance)
+// Given a inst proto.Instance this methid will update the instance in API with a PUT request
+func (m *Manager) pushInstanceInfo(inst proto.Instance) error {
+	// Subsystems will be ignored, don't send them
+	inst.Subsystems = make([]proto.Instance, 0)
+	uri := fmt.Sprintf("%s/%s", m.api.EntryLink("instances"), inst.UUID)
```
Comment (daniel-nichter) [inline instance/manager.go]: Same as above: ultimately this will go away when `sysconfig` is rewritten.

File: instance/repo.go [hunk 23]
```diff
@@ -18,257 +18,340 @@
...
+	configDir   string
+	api         pct.APIConnector
+	it          map[string]*proto.Instance
+	tree        *proto.Instance
+	treeVersion uint
+	mux         *sync.RWMutex
 }
 
+const (
+	INSTANCES_FILE     = "instance-tree.json" // relative to Repo.configDir
```
Comment (daniel-nichter) [inline instance/repo.go]: "system tree" not "instance tree" :-)

File: instance/repo.go [hunk 24]
```diff
@@ -18,257 +18,340 @@
...
+	INSTANCES_FILEMODE = 0660
+
+	// MYSQL_PREFIX an OS_PREFIX are the instance prefixes we need to validate
+	MYSQL_PREFIX = "mysql"
+	OS_PREFIX    = "os"
+)
+
+var UUID_RE, _ = regexp.Compile("^[0-9A-Fa-f]{32}$")
```
Comment (daniel-nichter) [inline instance/repo.go]: `MustCompile()`? (doesn't return error)

File: instance/repo.go [hunk 25]
```diff
@@ -18,257 +18,340 @@
...
 
-func (r *Repo) Init() error {
-	for service, _ := range proto.ExternalService {
-		if err := r.loadInstances(service); err != nil {
-			return fmt.Errorf("%s: %s", service, err)
+func isOSInstance(it proto.Instance) bool {
+	return it.Prefix == OS_PREFIX
+}
+
+func isMySQLInst(it proto.Instance) bool {
```
Comment (daniel-nichter) [inline instance/repo.go]: Be careful about mixing terms and abbreviations (i.e. consistency): in func above it's "Instance" but here it's "Inst". Imho this matters because if I search for "instance(" (case-insensitive) I can quickly find all instance-related funcs. Also, consistency = predictability for other devs = make their lives easier.

Comment (mpaolino) [inline instance/repo.go]: fixed, thanks

File: test/gocheck.go (hunk 9 above)
Comment (mpaolino) [inline test/gocheck.go]: this was fixed

**Comment by mpaolino** [comment]
Closing this PR, creating one from release-1.1
